tests/unit-tests/conversions.cpp

```diff
@@ -0,0 +1,62 @@
+#include "doctest/doctest.h"
+
+#include <ipr/impl>
+
+TEST_CASE("C++ Standard Conversions") {
+  using namespace ipr;
+  impl::Lexicon lexicon{};
+  impl::Module module{lexicon};
+  impl::Interface_unit unit{lexicon, module};
+
+  // static_cast<long long>(4);
+  //    Integral Promotion            (Promote)
+  lexicon.make_promote(
+    *lexicon.make_literal(lexicon.int_type(), "4"),
+    &lexicon.long_long_type()
+  );
+
+  // (float)2.2;
+  //    Floating-Point Conversion     (Demote)
+  lexicon.make_demote(
+    *lexicon.make_literal(lexicon.double_type(), "2.2"),
+    &lexicon.float_type()
+  );
+
+  // (volatile int)7;
+  //    CV-Qualifcation               (Qualification)
+  lexicon.make_qualification(
+    lexicon.get_qualified(Type_qualifier::Volatile, lexicon.int_type()),
+    *lexicon.make_literal(lexicon.double_type(), "2.2"),
+    &lexicon.int_type() // Type of the expr can differ from target type (see 7.2.2/2)
+  );
+
+  // int* const ptr = 0;
+  //    Pointer Conversion            (Coersion)
+  const auto& ptr_type = lexicon.get_qualified(
+    Type_qualifier::Const, lexicon.get_pointer(lexicon.int_type()));
+  auto& ptr = *unit.global_region()->declare_var(*lexicon.make_identifier("ptr"), ptr_type);
+  ptr.init = lexicon.make_coerce(
+    ptr_type,
+    *lexicon.make_literal(lexicon.int_type(), "0"),
+    &ptr_type
+  );
+
+  // if (ptr) double(6);
+  //    Boolean Conversion            (Coersion)
+  //    Lvalue-to-Rvalue              (Read)
+  //    Integral-Floating Conversion  (Coersion)
+  const auto& condition = *lexicon.make_coerce(
+    lexicon.bool_type(),
+    *lexicon.make_read(
+      *lexicon.make_id_expr(ptr),
+      &lexicon.get_pointer(lexicon.int_type()) // != ptr.type() (see 7.2.2/2)
+    ),
+    &lexicon.bool_type()
```

Comment:
I want to draw attention to this in particular for @GabrielDosReis, as I think I'm poorly informed. 

This `lexicon.bool_type` is being assigned to the `constraint` field of the `Coerce` class (inherited from `Expr`). I don't think I understand the purpose of a constraint, if this is as intended.

My expectation (based solely on the name) is that constraints are to be applied to type variables (such as uninstantiated template parameter types). This particular example has no type variables (they're all concrete), so I would expect there to be no constraints, so this use site appears incorrect (i.e., we should remove this parameter). 

Am I completely off base?

Reply:
Followed up offline. This code appears to be correct and my interpretation of `constraint` was incorrect. @GabrielDosReis can correct me if I misrepresent what he said:

At the time of IPR's creation, he was thinking of types like `int` as being predicates. The name `constraint` comes from that generalization of C++'s concept of type.